opal/mca/btl/tcp/btl_tcp_component.c

```diff
@@ -1299,45 +1338,156 @@ static void mca_btl_tcp_component_recv_handler(int sd, short flags, void* user)
+        if (ENOPROTOOPT == errno) {
+            sockopt = false;
+        } else {
+	    opal_output_verbose(20, opal_btl_base_framework.framework_output,
+				"Cannot get current recv timeout value of the socket"
+		                "Local_host:%s PID:%d",
+				opal_process_info.nodename, getpid());
+	    return;
+        }
+    } else {
+        tv.tv_sec = 2;
+        tv.tv_usec = 0;
+        if (0 != setsockopt(sd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv))) {
+            opal_output_verbose(20, opal_btl_base_framework.framework_output,
+				"Cannot set new recv timeout value of the socket"
+				"Local_host:%s PID:%d",
+				opal_process_info.nodename, getpid());
+           return;
+       }
+    }
 
     OBJ_RELEASE(event);
-
-    /* recv the process identifier */
-    retval = recv(sd, (char *)&guid, sizeof(guid), 0);
-    if(retval != sizeof(guid)) {
+    retval = mca_btl_tcp_recv_blocking(sd, (void *)&hs_msg, sizeof(hs_msg));
+    guid = hs_msg.guid;  
+
+    /* An unknown process attempted to connect to Open MPI via TCP.
+     * Open MPI uses a "magic" string to trivially verify that the connecting
+     * process is a fellow Open MPI process.  An MPI process accepted a TCP
+     * connection but did not receive the correct magic string.  This might
+     * indicate an Open MPI version mismatch between different MPI processes
+     * in the same job, or it may indicate that some other agent is
+     * mistakenly attempting to connect to Open MPI's TCP listening sockets.
+
+     * This attempted connection will be ignored; your MPI job may or may not
+     * continue properly.
+     */
+     if (sizeof(hs_msg) != retval) {
+         opal_output_verbose(20, opal_btl_base_framework.framework_output,
+                            "server did not receive entire connect ACK "
+                            "Local_host:%s PID:%d Role:%s String_received:%s Test_fail:%s",
+                            opal_process_info.nodename,
+                            getpid(), "server",
+                            (retval > 0) ? hs_msg.magic_id : "<nothing>", 
+                            "handshake message length");
+
+         /* The other side probably isn't OMPI, so just hang up */
+         CLOSE_THE_SOCKET(sd);
+         return;
+    }
+    if (0 != strncmp(hs_msg.magic_id, mca_btl_tcp_magic_id_string, len)) {
+        opal_output_verbose(20, opal_btl_base_framework.framework_output,
+                            "server did not receive right magic string. "
+                            "Local_host:%s PID:%d Role:%s String_received:%s Test_fail:%s",
+                            opal_process_info.nodename,
+                            getpid(), "server", hs_msg.magic_id,
+                            "string value");
+        /* The other side probably isn't OMPI, so just hang up */
         CLOSE_THE_SOCKET(sd);
         return;
     }
+    
+    if (sockopt) {
+       /* return the socket to normal */
```

Comment:
it may not be "normal".  I'd say "reset the RECVTIMEO option to its original state".